pkg/hubble/metrics/dynamic_flow_processor.go

```diff
@@ -0,0 +1,139 @@
+)
+
+// DynamicFlowProcessor represents instance of hubble exporter with dynamic
+// configuration reload.
+type DynamicFlowProcessor struct {
+	logger  logrus.FieldLogger
+	watcher *metricConfigWatcher
+	// Protects against deregistering metric handlers while ProcessFlow is executing, or concurrent config reloads.
+	mutex    lock.RWMutex
+	Metrics  []api.NamedHandler
+	registry *prometheus.Registry
+}
+
+// OnDecodedEvent distributes events across all managed exporters.
+func (d *DynamicFlowProcessor) OnDecodedFlow(ctx context.Context, flow *flowpb.Flow) (bool, error) {
+	select {
+	case <-ctx.Done():
+		return false, nil
+	default:
+	}
+
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
+
+	var errs error
+	if d.Metrics != nil {
+		errs = api.ExecuteAllProcessFlow(ctx, flow, &d.Metrics)
+	}
+
+	if errs != nil {
+		d.logger.WithError(errs).Error("Failed to ProcessFlow in metrics handler")
+	}
+	return false, errs
+}
+
+// NewDynamicFlowProcessor creates instance of dynamic hubble flow exporter.
+func NewDynamicFlowProcessor(reg *prometheus.Registry, logger logrus.FieldLogger, configFilePath string) *DynamicFlowProcessor {
+	dynamicFlowProcessor := &DynamicFlowProcessor{
+		logger:   logger,
+		registry: reg,
+	}
+	watcher := NewMetricConfigWatcher(configFilePath, dynamicFlowProcessor.onConfigReload)
+	dynamicFlowProcessor.watcher = watcher
+	return dynamicFlowProcessor
+}
+
+func (d *DynamicFlowProcessor) onConfigReload(ctx context.Context, hash uint64, config api.Config) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	var newHandlers []api.NamedHandler
+	metricNames := config.GetMetricNames()
+
+	curHandlerMap := make(map[string]*api.NamedHandler)
+	if d.Metrics != nil {
+		for _, m := range d.Metrics {
+			curHandlerMap[m.Name] = &m
+		}
+
+		configuredMetricNames := make(map[string]*api.MetricConfig)
+		for _, cm := range config.Metrics {
+			configuredMetricNames[cm.Name] = cm
+		}
+		// Unregister handlers not present in the new config.
+		// This needs to happen first to properly check for conflicting plugins later during registration.
+		for _, m := range d.Metrics {
+			if _, ok := configuredMetricNames[m.Name]; !ok {
+				h := curHandlerMap[m.Name]
+				err := h.Handler.Deinit(d.registry)
+				if err != nil {
+					d.logger.WithField("name", m.Name).WithError(err).Error("Deinit failed for handler")
+				}
+				delete(curHandlerMap, m.Name)
+			}
+		}
+	}
+
+	for _, cm := range config.Metrics {
+		// Existing handler matches new config entry:
+		//   no-op, if config unchanged;
+		//   update handler config, if changed.
+		if m, ok := curHandlerMap[cm.Name]; ok {
+			if reflect.DeepEqual(*m.MetricConfig, *cm) {
+				continue
+			} else {
+				if h, ok := curHandlerMap[cm.Name]; ok {
+					err := h.Handler.HandleConfigurationUpdate(cm)
+					if err != nil {
+						d.logger.WithField("name", cm.Name).WithError(err).Error("HandleConfigurationUpdate failed for handler")
+					}
+					h.MetricConfig = cm
+				}
```

Comment:
I'm not sure we should proceed if handling updates fails. Should we even proceed with the rest of the code? We probably don't want to update d.Metrics` if we get any errors, do we?

Reply:
Our validation functions rejects invalid config set reload attempts, eg. malformed params, or label changes - we take the all-or-nothing approach there. So when this code here executes and fails, that's currently due to unexpected Prometheus failures.

With errors only coming from Prometheus, we can't really have transactionality of config application, eg. for a this of 2 metrics [1: delete, 2: add]  if handler no.2 fails to register we can't cleanly undo no.1 having unregistered, since its metrics vectors will have been removed - we could only re-register it in an attempt to restore original state prior to applying the new config.

Since reverting to a clean state after partial failure cannot be achieved (can it?), I'd go for a best-effort approach of applying each config set that passed our validation.

Reply:
Right, well, I guess that's as good as it gets.